In [1]:
import numpy as np
import pandas as pd

In [2]:
raw_train = pd.read_csv("train.csv")

In [3]:
raw_train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [4]:
raw_train.shape

(891, 12)

In [5]:
raw_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [6]:
raw_train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [7]:
all_cols = list(raw_train)
num_data = [i for i in list(raw_train) if raw_train[i].dtypes != "object"]
cat_data = [i for i in list(raw_train) if raw_train[i].dtypes == "object"]
print(f"The numerical attributes are: {num_data}")

The numerical attributes are: ['PassengerId', 'Survived', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare']


### Convert sex to integer counterparts for more efficient data processing by model

#### convert pandas dataframe to numpy array to pass into the GenderTransformer class

In [8]:
raw_train_arr = raw_train.to_numpy()
raw_train_arr

array([[1, 0, 3, ..., 7.25, nan, 'S'],
       [2, 1, 1, ..., 71.2833, 'C85', 'C'],
       [3, 1, 3, ..., 7.925, nan, 'S'],
       ...,
       [889, 0, 3, ..., 23.45, nan, 'S'],
       [890, 1, 1, ..., 30.0, 'C148', 'C'],
       [891, 0, 3, ..., 7.75, nan, 'Q']], shape=(891, 12), dtype=object)

#### define the GenderTransform class and transform the data

In [9]:
from sklearn.base import BaseEstimator, TransformerMixin

class GenderTransformer(BaseEstimator, TransformerMixin):
    def __init__(self):
        pass
    def fit(self, X, y=None):
        return self
    def transform(self, X, y=None):
        X[:, 4]=np.where((X[:, 4] == 'female'), int(0), int(1))
        return X
    



In [10]:
sex_num = GenderTransformer()
raw_train_arr_sex_num = sex_num.transform(raw_train_arr)

In [11]:
corr_matrix = raw_train.corr(numeric_only=True)
corr_matrix["Survived"].sort_values(ascending=False)

Survived       1.000000
Fare           0.257307
Parch          0.081629
PassengerId   -0.005007
SibSp         -0.035322
Age           -0.077221
Pclass        -0.338481
Name: Survived, dtype: float64

#### reconsturct the pandas dataframe from the transformed numpy array

In [12]:
raw_train_sex_num = pd.DataFrame(raw_train_arr_sex_num, columns=all_cols)
raw_train_sex_num

PassengerId Survived Pclass  \
0             1        0      3   
1             2        1      1   
2             3        1      3   
3             4        1      1   
4             5        0      3   
..          ...      ...    ...   
886         887        0      2   
887         888        1      1   
888         889        0      3   
889         890        1      1   
890         891        0      3   

                                                  Name Sex   Age SibSp Parch  \
0                              Braund, Mr. Owen Harris   1  22.0     1     0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...   0  38.0     1     0   
2                               Heikkinen, Miss. Laina   0  26.0     0     0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)   0  35.0     1     0   
4                             Allen, Mr. William Henry   1  35.0     0     0   
..                                                 ...  ..   ...   ...   ...   
886                              Montvila, Rev. Juozas   1  27.0     0     0   
887                       Graham, Miss. Margaret Edith   0  19.0     0     0   
888           Johnston, Miss. Catherine Helen "Carrie"   0   NaN     1     2   
889                              Behr, Mr. Karl Howell   1  26.0     0     0   
890                                Dooley, Mr. Patrick   1  32.0     0     0   

               Ticket     Fare Cabin Embarked  
0           A/5 21171     7.25   NaN        S  
1            PC 17599  71.2833   C85        C  
2    STON/O2. 3101282    7.925   NaN        S  
3              113803     53.1  C123        S  
4              373450     8.05   NaN        S  
..                ...      ...   ...      ...  
886            211536     13.0   NaN        S  
887            112053     30.0   B42        S  
888        W./C. 6607    23.45   NaN        S  
889            111369     30.0  C148        C  
890            370376     7.75   NaN        Q  

[891 rows x 12 columns]

#### change to appropriate datatype wherever necessary

In [13]:
def revert_datatype(raw_train_sex_num):
    int_features_index = [0, 1, 2, 4, 6, 7]
    float_features_index = [5, 9]
    for i in int_features_index:
        raw_train_sex_num[all_cols[i]] = raw_train_sex_num[all_cols[i]].astype(np.int64)
    for i in float_features_index:
        raw_train_sex_num[all_cols[i]] = raw_train_sex_num[all_cols[i]].astype(float) 
revert_datatype(raw_train_sex_num)

#### construct the correlation matrix for the transformed data

In [14]:
corr_mat = raw_train_sex_num.corr(numeric_only=True)
corr_mat["Survived"].sort_values(ascending=True)

Sex           -0.543351
Pclass        -0.338481
Age           -0.077221
SibSp         -0.035322
PassengerId   -0.005007
Parch          0.081629
Fare           0.257307
Survived       1.000000
Name: Survived, dtype: float64

### Transform the Embarked column with OneHotEncoder

#### The approach I am using is:
- Firstly clean the data by filling the nan columns
- To fill the nan values i am using the most frequent value of the column 

In [15]:
raw_train_sex_num['Embarked']

0      S
1      C
2      S
3      S
4      S
      ..
886    S
887    S
888    S
889    C
890    Q
Name: Embarked, Length: 891, dtype: object

In [16]:
raw_train_sex_num['Embarked'].value_counts().idxmax()

'S'

#### class to fill the nan values present in Embarked column

In [17]:
from sklearn.base import BaseEstimator, TransformerMixin

class FillnaEmbarked(BaseEstimator, TransformerMixin):
    def __init__(self):
        pass
    def fit(self, X, y=None):
        return self
    def transform(self, X, y=None):
        X["Embarked"] = X["Embarked"].fillna(X["Embarked"].value_counts().idxmax())
        return X.to_numpy()

In [18]:
fillna_embark = FillnaEmbarked()
raw_train_embark_arr = fillna_embark.transform(raw_train_sex_num)
raw_train_embark_filled = pd.DataFrame(raw_train_embark_arr, columns=all_cols)
raw_train_embark_filled

PassengerId Survived Pclass  \
0             1        0      3   
1             2        1      1   
2             3        1      3   
3             4        1      1   
4             5        0      3   
..          ...      ...    ...   
886         887        0      2   
887         888        1      1   
888         889        0      3   
889         890        1      1   
890         891        0      3   

                                                  Name Sex   Age SibSp Parch  \
0                              Braund, Mr. Owen Harris   1  22.0     1     0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...   0  38.0     1     0   
2                               Heikkinen, Miss. Laina   0  26.0     0     0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)   0  35.0     1     0   
4                             Allen, Mr. William Henry   1  35.0     0     0   
..                                                 ...  ..   ...   ...   ...   
886                              Montvila, Rev. Juozas   1  27.0     0     0   
887                       Graham, Miss. Margaret Edith   0  19.0     0     0   
888           Johnston, Miss. Catherine Helen "Carrie"   0   NaN     1     2   
889                              Behr, Mr. Karl Howell   1  26.0     0     0   
890                                Dooley, Mr. Patrick   1  32.0     0     0   

               Ticket     Fare Cabin Embarked  
0           A/5 21171     7.25   NaN        S  
1            PC 17599  71.2833   C85        C  
2    STON/O2. 3101282    7.925   NaN        S  
3              113803     53.1  C123        S  
4              373450     8.05   NaN        S  
..                ...      ...   ...      ...  
886            211536     13.0   NaN        S  
887            112053     30.0   B42        S  
888        W./C. 6607    23.45   NaN        S  
889            111369     30.0  C148        C  
890            370376     7.75   NaN        Q  

[891 rows x 12 columns]

#### restore the data to proper datatype

In [19]:
revert_datatype(raw_train_embark_filled)

#### seperate the embarked values using OneHotEncoder for ease of model

In [21]:
from sklearn.preprocessing import OneHotEncoder
one_hot_encoder = OneHotEncoder()
encoded_embarked = one_hot_encoder.fit_transform(raw_train_sex_num[["Embarked"]])

In [22]:
encoded_embarked.toarray()

array([[0., 0., 1.],
       [1., 0., 0.],
       [0., 0., 1.],
       ...,
       [0., 0., 1.],
       [1., 0., 0.],
       [0., 1., 0.]], shape=(891, 3))

#### Steps to add the OneHotEncoded values to the original dataset
- we get a 891, 3 shapped array which at first we convert to a single dimention array
- the first index, second and third indeces of the array indicated C, Q and S values respectively
- so in the 1d array if we traverse with step size of 3 we will get C, Q and S values respectively
- created a dictionay with keys: C, Q, S and add the data from the 1d array three at one iteration
- create a pandas dataframe out of the dictionay
- join the existing dataframe and the new formed dataframe
- drop the embarked column present in the original dataset

In [23]:
encoded_embarked_arr = encoded_embarked.toarray().reshape(2673)

In [24]:
onehot_categories = []
for i in one_hot_encoder.categories_:
    onehot_categories.extend(i)
onehot_categories

['C', 'Q', 'S']

In [25]:
encoded_dict = {
    onehot_categories[0] : [],
    onehot_categories[1]: [],
    onehot_categories[2]: [],
}
for i in range(0, len(encoded_embarked_arr)-2, 3):
    encoded_dict['C'].append(int(encoded_embarked_arr[i+0]))
    encoded_dict['Q'].append(int(encoded_embarked_arr[i+1]))
    encoded_dict['S'].append(int(encoded_embarked_arr[i+2]))

In [26]:

raw_train_embark_filled = raw_train_embark_filled.join(pd.DataFrame(encoded_dict))

In [27]:
raw_train_embark_filled.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    int64  
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     891 non-null    object 
 12  C            891 non-null    int64  
 13  Q            891 non-null    int64  
 14  S            891 non-null    int64  
dtypes: float64(2), int64(9), object(4)
memory usage: 104.5+ KB


In [30]:
raw_train_embark_filled.drop(columns=["Embarked"], inplace=True)
raw_train_embark_filled

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name  Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    1  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...    0  38.0      1   
2                               Heikkinen, Miss. Laina    0  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)    0  35.0      1   
4                             Allen, Mr. William Henry    1  35.0      0   
..                                                 ...  ...   ...    ...   
886                              Montvila, Rev. Juozas    1  27.0      0   
887                       Graham, Miss. Margaret Edith    0  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"    0   NaN      1   
889                              Behr, Mr. Karl Howell    1  26.0      0   
890                                Dooley, Mr. Patrick    1  32.0      0   

     Parch            Ticket     Fare Cabin  C  Q  S  
0        0         A/5 21171   7.2500   NaN  0  0  1  
1        0          PC 17599  71.2833   C85  1  0  0  
2        0  STON/O2. 3101282   7.9250   NaN  0  0  1  
3        0            113803  53.1000  C123  0  0  1  
4        0            373450   8.0500   NaN  0  0  1  
..     ...               ...      ...   ... .. .. ..  
886      0            211536  13.0000   NaN  0  0  1  
887      0            112053  30.0000   B42  0  0  1  
888      2        W./C. 6607  23.4500   NaN  0  0  1  
889      0            111369  30.0000  C148  1  0  0  
890      0            370376   7.7500   NaN  0  1  0  

[891 rows x 14 columns]

### Split the data into test set and train set
#### as we see in the last correlation matrix that the survived column has very high dependency on the sex feature so we are dividing the dataset based on the sex parameter

#### split the data in a ratio of 7:3

In [32]:
from sklearn.model_selection import StratifiedShuffleSplit
train_test_split = StratifiedShuffleSplit(n_splits=1, test_size=0.3, random_state=42)
for train_index, test_index in train_test_split.split(raw_train_embark_filled, raw_train_embark_filled["Sex"]):
    X_train = raw_train_embark_filled.loc[train_index]
    X_test = raw_train_embark_filled.loc[test_index]

#### seperate the data from the labels in both the train set and the test set

In [36]:
y_train = X_train["Survived"]
y_test = X_test["Survived"]
X_train.drop(columns=["Survived"], inplace=True)
X_test.drop(columns=["Survived"], inplace=True)

In [37]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 623 entries, 97 to 421
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  623 non-null    int64  
 1   Pclass       623 non-null    int64  
 2   Name         623 non-null    object 
 3   Sex          623 non-null    int64  
 4   Age          507 non-null    float64
 5   SibSp        623 non-null    int64  
 6   Parch        623 non-null    int64  
 7   Ticket       623 non-null    object 
 8   Fare         623 non-null    float64
 9   Cabin        142 non-null    object 
 10  C            623 non-null    int64  
 11  Q            623 non-null    int64  
 12  S            623 non-null    int64  
dtypes: float64(2), int64(8), object(3)
memory usage: 68.1+ KB


In [38]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
Index: 268 entries, 154 to 563
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  268 non-null    int64  
 1   Pclass       268 non-null    int64  
 2   Name         268 non-null    object 
 3   Sex          268 non-null    int64  
 4   Age          207 non-null    float64
 5   SibSp        268 non-null    int64  
 6   Parch        268 non-null    int64  
 7   Ticket       268 non-null    object 
 8   Fare         268 non-null    float64
 9   Cabin        62 non-null     object 
 10  C            268 non-null    int64  
 11  Q            268 non-null    int64  
 12  S            268 non-null    int64  
dtypes: float64(2), int64(8), object(3)
memory usage: 29.3+ KB


### Tackle the null values present in the sex column
#### its not likely to drop the rows which has null values in the sex column because it has very high dependency on the survived column